### ISSUE TITLE
ComponentBank is bad design - look into a central event queue.

### ISSUE THREAD

**Issue by a-day-old-bagel** [a-day-old-bagel/DualityEngine#20]
So it's bothered me for a while: As I was building the skeleton for the engine, I decided that the way to give all systems access to the same game data (the components) would be to construct a singleton called Bank.  Well, that flies in the face of some game engine wisdom (such as described on [this](http://www.gamedev.net/page/resources/_/technical/game-programming/making-a-game-engine-core-design-principles-r3210) page).
I felt dirty as I wrote it, and downright filthy when I later began the inevitable decline into the practice of adding all kinds of junk to the bank whenever there was some non-component data that needed to be shared. The whole bank now reeks of code smell, in my opinion.  I feel like it's going to lead to problems later on, and I'm trying to think of more elegant ways to have all the systems access component data.

So, the most obvious approach (which I originally avoided due to some notion that it would be slow, which it might) is a [central event queue](http://gameprogrammingpatterns.com/event-queue.html).  The event submission mechanism will still have to be global, of course, taking the volatile place of the bank.  But the bank could be kept in Game, at the top level, where also the event handling loop could be implemented.  Events like "add [xyz] transform to entity [id]" could be sent in from physics systems, and the bank itself could be modified only from Game, using the main thread.  The main thread currently is unused (it just waits to finish), so there's no problem there. This would make future parallelism much easier, potentially remaining lock-free.

Perhaps each system could also contain an individual event queue, to which Game (and only Game) could send events.  Currently each system loops as fast as it can, which is completely asynchronous, but may also waste CPU and GPU time.  Consider the case where the physics thread updates the game, then the graphics thread draws everything _twice_ before the physics thread updates the second time.  If the main thread were to send an "update" event to the physics system and a "draw" event to the rendering systems every n milliseconds, both systems could still do their work in parallel, but wouldn't get needlessly ahead of each other so much.  The sound system would be another natural place for an event queue, and the scripting system already uses one.  These system-specific event queues aren't required for the central event queue to work, but I just thought that while we're at it, we might as well...

Anyway, ideas are appreciated.  I might start a branch to see if I can throw together such a queue and see how well it works.

**Comment by natevm** [comment]
I don't think we should keep bank around, since we're flexible now, but won't be in the future. Our alternative should be able to cleanly replace the bank solution.

Could you create an entity relationship diagram of the revised queue system, and send it to me for peer review/training? You could create multiple designs and we can quickly discuss what ones are better than others.

**Comment by natevm** [comment]
The entity relation diagram could be drawn on a napkin. Nothing fancy.

**Comment by a-day-old-bagel** [comment]
The problem is the need to access shared non-component data.  For example, there may be many camera components scattered throughout all the entities in the scene, but only one of them is going to be the _active_ camera, meaning that it's the one through which the user is currently looking.  Right now an id is stored in bank called activeCameraId or something.  More than one system needs to know that id, unless we want to have to search through all camera components from each system each frame to find the right one (we don't.)  A compromise might be to keep something like the bank around, but allow only read operations (getters) to be public, forcing writing to be done through the event queue.  I don't know - this is a difficult problem.  I want to keep performance as our top objective, while at the same time having a more elegant design.

**Comment by a-day-old-bagel** [comment]
Ok, I think I have an idea of how to do this.  Note that while looking at the red/green arrows, the master thread consumes red and produces green, while the system threads consume green and produce red.  As long as the red event flux is greater than the green event flux, we only have to regulate (handle over/underflow for the main thread (central event queue).  That's another reason to ditch the idea of update/draw events (see below).
(flux meaning flow through an area)
![CEQ design idea](http://i.imgur.com/x9gzsJs.png)

**Comment by a-day-old-bagel** [comment]
ok, I'm going to try moodycamel's implementation of the [concurrent queue](https://github.com/cameron314/concurrentqueue) for the ceq.
